packages/sdk38/src/secp256k1wallet.ts

```diff
@@ -0,0 +1,332 @@
+import { fromBase64, fromHex, fromUtf8, toAscii, toBase64, toHex, toUtf8 } from "@cosmjs/encoding";
+import { assert, isNonNullObject } from "@cosmjs/utils";
+
+import { rawSecp256k1PubkeyToAddress } from "./address";
+import { encodeSecp256k1Signature } from "./signature";
+import { StdSignature } from "./types";
+import {
+  AccountData,
+  Algo,
+  executeKdf,
+  KdfConfiguration,
+  makeCosmoshubPath,
+  OfflineSigner,
+  prehash,
+  PrehashType,
+} from "./wallet";
+
+const serializationTypeV1 = "secp256k1wallet-v1";
+
+/**
+ * A fixed salt is chosen to archive a deterministic password to key derivation.
+ * This reduces the scope of a potential rainbow attack to all CosmJS users.
+ * Must be 16 bytes due to implementation limitations.
+ */
+const cosmjsSalt = toAscii("The CosmJS salt.");
+
+/**
+ * A KDF configuration that is not very strong but can be used on the main thread.
+ * It takes about 1 second in Node.js 12.15 and should have similar runtimes in other modern Wasm hosts.
+ */
+const basicPasswordHashingOptions: KdfConfiguration = {
+  algorithm: "argon2id",
+  params: {
+    outputLength: 32,
+    opsLimit: 20,
+    memLimitKib: 12 * 1024,
+  },
+};
+
+const algorithmIdXchacha20poly1305Ietf = "xchacha20poly1305-ietf";
+
+/**
+ * This interface describes a JSON object holding the encrypted wallet and the meta data.
+ * All fields in here must be JSON types.
+ */
+export interface Secp256k1WalletSerialization {
+  /** A format+version identifier for this serialization format */
+  readonly type: string;
+  /** Information about the key derivation function (i.e. password to encrytion key) */
+  readonly kdf: KdfConfiguration;
+  /** Information about the symmetric encryption */
+  readonly encryption: {
+    /**
+     * An algorithm identifier, such as "xchacha20poly1305-ietf".
+     */
+    readonly algorithm: string;
+    /** A map of algorithm-specific parameters */
+    readonly params: Record<string, unknown>;
+  };
+  /** An instance of Secp256k1WalletData, which is stringified, encrypted and base64 encoded. */
+  readonly data: string;
+}
+
+/**
+ * The data of a wallet serialization that is encrypted.
+ * All fields in here must be JSON types.
+ */
+export interface Secp256k1WalletData {
+  readonly mnemonic: string;
+  readonly accounts: ReadonlyArray<{
+    readonly algo: string;
+    readonly hdPath: string;
+    readonly prefix: string;
+  }>;
+}
+
+function extractKdfConfigurationV1(document: any): KdfConfiguration {
+  return document.kdf;
+}
+
+export function extractKdfConfiguration(serialization: string): KdfConfiguration {
+  const root = JSON.parse(serialization);
+  if (!isNonNullObject(root)) throw new Error("Root document is not an object.");
+
+  switch ((root as any).type) {
+    case serializationTypeV1:
+      return extractKdfConfigurationV1(root);
+    default:
+      throw new Error("Unsupported serialization type");
+  }
+}
+
+export class Secp256k1Wallet implements OfflineSigner {
```

Comment:
Okay, so one class for the basic usage (provide mnemonic) and for the complex usage (encrypted storage). Let me see if this does make sense to combine.

Reply:
It turned out that the existing use cases (in-memory wallet and signing) don't become more complicated. We can even add a very simple `generate`. The interfaces only become non-trivial when dealing with serialization/deserialization.